Sources/OuterCore/IRVerifier.swift

```diff
@@ -0,0 +1,165 @@
+//
+//  IRVerifier.swift
+//  OuterCore
+//
+//  Created by Harlan Haskins on 1/28/18.
+//
+
+import Foundation
+import Lithosphere
+
+extension Diagnostic.Message {
+  static func unknownType(_ type: Type) -> Diagnostic.Message {
+    return .init(.error, "Graph IR: unknown type '\(name(for: type))'")
+  }
+  static func continuationHasNoCall(
+    _ continuation: Continuation) -> Diagnostic.Message {
+    return .init(.error,
+                 """
+                 Graph IR: continuation '\(continuation.name)' has no call
+                 """)
+  }
+  static func callingNonFunction(_ type: Type) -> Diagnostic.Message {
+    return .init(.error,
+                 """
+                 Graph IR: attempt to call non-function type \
+                 '\(name(for: type))'
+                 """)
+  }
+  static func arityMismatch(function: Value, expected: Int,
+                            got: Int) -> Diagnostic.Message {
+    let qualifier = expected > got ? "few" : "many"
+    let kind = function is Continuation ? "function" : "value"
+    return .init(.error,
+                 """
+                 Graph IR: too \(qualifier) arguments to Graph IR \(kind) \
+                 '\(function.name)'; expected \(expected), got \(got)
+                 """)
+  }
+  static func typeMismatch(expected: Type, got: Type) -> Diagnostic.Message {
+    return .init(.error,
+                 """
+                 Graph IR: type mismatch (expected '\(name(for: expected))', \
+                 got '\(name(for: got))')
+                 """)
+  }
+
+  static func unsubstitutedType(
+    _ type: ParameterizedType) -> Diagnostic.Message {
+    return .init(.error,
+                 """
+                 Graph IR: type \(name(for: type)) must have all parameters \
+                 substituted
+                 """)
+  }
+}
+
+public final class IRVerifier {
+  let module: Module
+  let engine: DiagnosticEngine
+
+  var currentScopedValues = Set<Value>()
+
+  public init(module: Module, engine: DiagnosticEngine) {
+    self.module = module
+    self.engine = engine
+  }
+
+  func withScope<T>(_ actions: () -> T) -> T {
```

Comment:
GraphIR should have a flat scope.  All references are done by qualified name.

Reply:
If you're not getting qualified names by this point, ScopeCheck is broken.  (And ScopeCheck Is Usually Broken)

Reply:
I really don’t think we should verify via qualified name, at all. GraphIR is a graph, names don’t mean anything here. Names are only for debugging.

Reply:
Names of functions and projections is what I mean.